### PR TITLE
Tobin gh1869 ds shared ptr

### LINKED ISSUE #1869: DescriptorSets: Use shared_ptr for DescriptorSetLayout data instead of copies
DescriptorSets need access to DSLayout state, which was handled via pointers (problematic if DSLayouts were destroyed) but is now handled by making copies of the DSLayout data for each descriptorSet.  This could be better managed by using shared_ptrs.

### PR DESCRIPTION
**Pull request by tobine**
Fixes #1869 

Migrate DescriptorSetLayout references to use shared_ptr.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -4532,8 +4532,8 @@ static bool PreCallValidateCreateDescriptorSetLayout(layer_data *dev_data, const
 
 static void PostCallRecordCreateDescriptorSetLayout(layer_data *dev_data, const VkDescriptorSetLayoutCreateInfo *create_info,
                                                     VkDescriptorSetLayout set_layout) {
-    dev_data->descriptorSetLayoutMap[set_layout] = std::unique_ptr<cvdescriptorset::DescriptorSetLayout>(
-        new cvdescriptorset::DescriptorSetLayout(create_info, set_layout));
+    dev_data->descriptorSetLayoutMap[set_layout] =
+        std::shared_ptr<cvdescriptorset::DescriptorSetLayout>(new cvdescriptorset::DescriptorSetLayout(create_info, set_layout));
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: std::make_shared!

File: layers/descriptor_sets.cpp [hunk 2]
```diff
@@ -312,21 +313,21 @@ cvdescriptorset::AllocateDescriptorSetsData::AllocateDescriptorSetsData(uint32_t
     : required_descriptors_by_type{}, layout_nodes(count, nullptr) {}
 
 cvdescriptorset::DescriptorSet::DescriptorSet(const VkDescriptorSet set, const VkDescriptorPool pool,
-                                              const DescriptorSetLayout *layout, const layer_data *dev_data)
+                                              std::shared_ptr<DescriptorSetLayout> layout, const layer_data *dev_data)
```
Comment (chrisforbes) [inline layers/descriptor_sets.cpp]: Consider passing shared_ptr by ref when you don't need the temporary to be an owner (but you will make a copy of it elsewhere)

File: layers/descriptor_sets.cpp [hunk 3]
```diff
@@ -389,8 +390,8 @@ static std::string string_descriptor_req_view_type(descriptor_req req) {
 }
 
 // Is this sets underlying layout compatible with passed in layout according to "Pipeline Layout Compatibility" in spec?
-bool cvdescriptorset::DescriptorSet::IsCompatible(const DescriptorSetLayout *layout, std::string *error) const {
-    return layout->IsCompatible(&p_layout_, error);
+bool cvdescriptorset::DescriptorSet::IsCompatible(std::shared_ptr<DescriptorSetLayout> const layout, std::string *error) const {
```
Comment (chrisforbes) [inline layers/descriptor_sets.cpp]: This should be a non-owning ptr.

File: layers/descriptor_sets.h [hunk 4]
```diff
@@ -110,7 +110,7 @@ class DescriptorSetLayout {
     bool HasBinding(const uint32_t binding) const { return binding_to_index_map_.count(binding) > 0; };
     // Return true if this layout is compatible with passed in layout,
     //   else return false and update error_msg with description of incompatibility
-    bool IsCompatible(const DescriptorSetLayout *, std::string *) const;
+    bool IsCompatible(std::shared_ptr<DescriptorSetLayout> const, std::string *) const;
```
Comment (chrisforbes) [inline layers/descriptor_sets.h]: Non-owning!

File: layers/descriptor_sets.h [hunk 5]
```diff
@@ -370,7 +363,7 @@ class DescriptorSet : public BASE_NODE {
     // Perform a CopyUpdate whose contents were just validated using ValidateCopyUpdate
     void PerformCopyUpdate(const VkCopyDescriptorSet *, const DescriptorSet *);
 
-    const DescriptorSetLayout *GetLayout() const { return &p_layout_; };
+    std::shared_ptr<DescriptorSetLayout> const GetLayout() const { return p_layout_; };
```
Comment (chrisforbes) [inline layers/descriptor_sets.h]: This was previously a ptr to a const DescriptorSetLayout. Now it's a const shared_ptr to a (nonconst) DescriptorSetLayout. Are you sure that's what you want?

Same in a few other places.

**Comment by tobine** [comment]
@chrisforbes yep, thanks, I'll clean these up

**Pull request merged**